### DISCUSSION TITLE
Error with boolean_difference

### DISCUSSION THREAD

**Discussion by D-Navajas** [pyvista/pyvista#2379] [Q&A] [answered]
Hi all, I hope you are doing well.

I have found an error I can't understand. I am trying to do a boolean difference between a plane and a cylinder, but the code doesn't finish running.  I have found this problem in several of my surfaces, but here is an example:

```
import pyvista as pv
import numpy as np

radius = 0.03795
center = np.array([0.08233, 0.1435, 0.09058])
p2 = np.array([0.02233, 0.1435, 0.09058])

direction = p2 - center
height = np.linalg.norm(direction)

centroid = center + direction/2

cutter = pv.Cylinder(
    center=centroid, direction=direction, radius=radius,
    height=height, resolution=100, capping=False
).triangulate()

points = np.array([[0.05233, 0.17432, 0.06844],
                 [0.05233, 0.17897, 0.06509],
                 [0.05233, 0.17782, 0.10677],
                 [0.05233, 0.20435, 0.11079]])
faces = np.array([4, 0, 1, 3, 2])
plane = pv.PolyData(points, faces).triangulate()

cut = plane.boolean_difference(cutter)
cut.plot()

pl = pv.Plotter()
pl.add_mesh(cut, show_edges=True)
pl.add_mesh(cutter, style="wireframe")
pl.show()
```

I hope you can help me find a solution, thanks in advance

**Comment by darikg** [comment, marked as answer]
I don't think the boolean operations are well-defined for bounded manifolds like the plane.

You might be able to achieve something similar with the [vtkIntersectionPolyDataFilter](vtkIntersectionPolyDataFilter), which isn't implemented in pyvista yet:

```python
# The first time I tried this it crashed, I think because two of the points of the plane sit exactly on cylinder faces.
# Moved them manually here 
points = np.array([[0.05233, 0.15, 0.06844],
                 [0.05233, 0.17897, 0.06509],
                 [0.05233, 0.15782, 0.10677],
                 [0.05233, 0.20435, 0.11079]])

...

from vtk import vtkIntersectionPolyDataFilter
from pyvista import Plotter
from pyvista.core.filters import _get_output

ifilt = vtkIntersectionPolyDataFilter()
ifilt.SetInputData(0, plane)
ifilt.SetInputData(1, cutter)
ifilt.Update()

plane1 = _get_output(ifilt, oport=1)
cutter1 = _get_output(ifilt, oport=2)

p = Plotter()
p.add_mesh(plane1, show_edges=True)
p.add_mesh(cutter1, show_edges=True)
p.show()
```

![image](https://user-images.githubusercontent.com/6875882/160196389-0568d790-6c15-4f1c-89b8-d6d301c77435.png)

**Reply by darikg** [reply]
just occurred to be that maybe the boolean difference will do exactly what you wanted after all, it's just failing because of the plane vertices coplanar on the cylinder surface. I'm pretty sure the boolean filters apply vtkIntersectionPolyDataFilter as a first pass, so if that fails so will the boolean.

**Reply by D-Navajas** [reply]
It was as you said, the points in the plane were on the edge of the cylinder. By adding 1e-5 to the radius of the cylinder the problem was solved. But I have found another combination that crashes, but this time the cylinder doesn't touch the vertices:
![2022-03-26_12h01_48](https://user-images.githubusercontent.com/10737095/160237164-46ec3d38-c949-4460-bd0d-6de1c80c9fc2.png)
```
import pyvista as pv
import numpy as np

radius = 0.046
center = np.array([0.24633, 0.07364, 0.15397])
p2 = np.array([0.29633, 0.07364, 0.15397])

direction = p2 - center
height = np.linalg.norm(direction)*2

centroid = center + direction/2

cutter = pv.Cylinder(
    center=centroid, direction=direction, radius=radius,
    height=height, resolution=100, capping=False
).triangulate()

points = np.array([[0.22194, 0.06364, 0.049],
                 [0.22194, 0.10014, 0.049],
                 [0.24383, 0.124, 0.15397],
                 [0.24383, 0.0225, 0.15397]])
faces = np.array([4, 0, 1, 2, 3,])
plane = pv.PolyData(points, faces).triangulate()

cut = plane.boolean_difference(cutter)
mask = cut.cell_data["CellSource"]
surfaces = cut.remove_cells(mask==1)

pl = pv.Plotter()
pl.add_mesh(surfaces, show_edges=True)
pl.add_mesh(cutter, style="wireframe")
pl.show()
```

I am a beginner with vtk and pyvista, so I don't know if there is a problem with these surfaces. I have also tried using ´clip_surfaces´, but the result wasn't really precise in the cases it even detected the intersection. What should I do? Thanks

**Reply by darikg** [reply]
Looks like it works if you add a small amount of jitter to the plane, i.e. `from numpy.random import uniform; points += 0.001 * uniform(0, 1, (4, 3))`. I would guess there is an exact edge-edge intersection and the vtk intersection filter only handles triangles in general position.

**Reply by D-Navajas** [reply]
Thanks! That worked perfectly! 
As I need to do some postprocessing, is there a easy way to remove those extra points after the cut?
